### IMPORT ALL LIBS

In [1]:
pip install nltk

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd#data manipulation,loading

import numpy as np#numerical and mathematical problem solving

import matplotlib as plt#data visualization

import re#pattern matching in string and transform and clean text

#nltk is natural language toolkit

from nltk.corpus import stopwords#words like what when ,it does not add values 

from  nltk.stem.porter import PorterStemmer#reduces words to rootform like running to run

from sklearn.feature_extraction.text import TfidfVectorizer#This is used to convert a collection of raw documents
#into a matrix of TF-IDF (Term Frequency-Inverse Document Frequency) features. 
#It reflects how important a word is to a document in a collection

In [3]:
from sklearn.linear_model import LogisticRegression#to train the model

In [4]:
from sklearn.model_selection import train_test_split#splitting to train and test

In [5]:
from sklearn.metrics import accuracy_score#to calculate the accuracy of model

In [6]:
import nltk

In [7]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\LENOVO\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

### Printing the words that can be skip because it won't play any important role for the model

In [8]:
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

### 1.	Data Collection and Exploration:

In [9]:
ls

 Volume in drive C is Windows-SSD
 Volume Serial Number is 2E3D-90D7

 Directory of C:\Users\LENOVO\Fake news prediction

08/19/2024  04:38 AM    <DIR>          .
08/19/2024  03:07 AM    <DIR>          ..
08/19/2024  03:08 AM    <DIR>          .ipynb_checkpoints
08/19/2024  04:38 AM            71,305 Fake_news_pred.ipynb
08/19/2024  03:09 AM            41,609 submit.csv
08/19/2024  03:09 AM        25,144,581 test.csv
08/19/2024  03:09 AM        98,628,550 train.csv
               4 File(s)    123,886,045 bytes
               3 Dir(s)  158,854,885,376 bytes free


### Loading the csv file to variable to dataset using pandas

In [10]:
dataset=pd.read_csv('train.csv')

In [11]:
dataset.head(12)#first 12 rows of the dataset

id                                              title  \
0    0  House Dem Aide: We Didn’t Even See Comey’s Let...   
1    1  FLYNN: Hillary Clinton, Big Woman on Campus - ...   
2    2                  Why the Truth Might Get You Fired   
3    3  15 Civilians Killed In Single US Airstrike Hav...   
4    4  Iranian woman jailed for fictional unpublished...   
5    5  Jackie Mason: Hollywood Would Love Trump if He...   
6    6  Life: Life Of Luxury: Elton John’s 6 Favorite ...   
7    7  Benoît Hamon Wins French Socialist Party’s Pre...   
8    8  Excerpts From a Draft Script for Donald Trump’...   
9    9  A Back-Channel Plan for Ukraine and Russia, Co...   
10  10  Obama’s Organizing for Action Partners with So...   
11  11  BBC Comedy Sketch "Real Housewives of ISIS" Ca...   

                          author  \
0                  Darrell Lucus   
1                Daniel J. Flynn   
2             Consortiumnews.com   
3                Jessica Purkiss   
4                 Howard Portnoy   
5                Daniel Nussbaum   
6                            NaN   
7                Alissa J. Rubin   
8                            NaN   
9   Megan Twohey and Scott Shane   
10                   Aaron Klein   
11               Chris Tomlinson   

                                                 text  label  
0   House Dem Aide: We Didn’t Even See Comey’s Let...      1  
1   Ever get the feeling your life circles the rou...      0  
2   Why the Truth Might Get You Fired October 29, ...      1  
3   Videos 15 Civilians Killed In Single US Airstr...      1  
4   Print \nAn Iranian woman has been sentenced to...      1  
5   In these trying times, Jackie Mason is the Voi...      0  
6   Ever wonder how Britain’s most iconic pop pian...      1  
7   PARIS  —   France chose an idealistic, traditi...      0  
8   Donald J. Trump is scheduled to make a highly ...      0  
9   A week before Michael T. Flynn resigned as nat...      0  
10  Organizing for Action, the activist group that...      0  
11  The BBC produced spoof on the “Real Housewives...      0

In [12]:
dataset.shape

(20800, 5)

### There are 20800 rows and 5 columns

In [13]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20800 entries, 0 to 20799
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      20800 non-null  int64 
 1   title   20242 non-null  object
 2   author  18843 non-null  object
 3   text    20761 non-null  object
 4   label   20800 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 812.6+ KB


### Let's see if values are null or not

In [14]:
dataset.isnull().sum()

id           0
title      558
author    1957
text        39
label        0
dtype: int64

### So there are in 3 columns

In [15]:

missing=dataset.isnull().sum().sum()

### Percentage of missing values

In [16]:
(missing/dataset.shape[0])*100

12.278846153846153

#### 12 percent  of value of total valuel are missing in the dataset as compared to rows

#### From initial exploration: We understand the data with data.head() and found that it has 5 columns where 4 are input and 1 i.e label is output

#### Structure:It tells us that it has some id , title of the news , its author ,its text or news  and from that info it will say its true news or false

##### Output:0 states as reliable news and 1 as unreliable news

#### Datatypes:It has 20800 rows of data,except id the inputs are all categorical data but the label output is boolean

#### Potential issues: It has 12.3 percent about missing value , so we have to fill according to the data

### 2.	Data Cleaning and Transformation

#### Filling Null values

#### As the data is very large , and understanding the context of the dataset it is good to replace the null value with null string

##### Null string helps the dataset to not overfit the data

In [17]:
dataset=dataset.fillna('')

In [18]:
dataset.isnull().sum()#all the null values are filled with empty string

id        0
title     0
author    0
text      0
label     0
dtype: int64

In [19]:
dataset.head()

id                                              title              author  \
0   0  House Dem Aide: We Didn’t Even See Comey’s Let...       Darrell Lucus   
1   1  FLYNN: Hillary Clinton, Big Woman on Campus - ...     Daniel J. Flynn   
2   2                  Why the Truth Might Get You Fired  Consortiumnews.com   
3   3  15 Civilians Killed In Single US Airstrike Hav...     Jessica Purkiss   
4   4  Iranian woman jailed for fictional unpublished...      Howard Portnoy   

                                                text  label  
0  House Dem Aide: We Didn’t Even See Comey’s Let...      1  
1  Ever get the feeling your life circles the rou...      0  
2  Why the Truth Might Get You Fired October 29, ...      1  
3  Videos 15 Civilians Killed In Single US Airstr...      1  
4  Print \nAn Iranian woman has been sentenced to...      1

#### We can mix title and author so it reduces the column number

In [20]:
dataset['content']=dataset['author'] + ' ' + dataset['title']

In [21]:
dataset

id                                              title  \
0          0  House Dem Aide: We Didn’t Even See Comey’s Let...   
1          1  FLYNN: Hillary Clinton, Big Woman on Campus - ...   
2          2                  Why the Truth Might Get You Fired   
3          3  15 Civilians Killed In Single US Airstrike Hav...   
4          4  Iranian woman jailed for fictional unpublished...   
...      ...                                                ...   
20795  20795  Rapper T.I.: Trump a ’Poster Child For White S...   
20796  20796  N.F.L. Playoffs: Schedule, Matchups and Odds -...   
20797  20797  Macy’s Is Said to Receive Takeover Approach by...   
20798  20798  NATO, Russia To Hold Parallel Exercises In Bal...   
20799  20799                          What Keeps the F-35 Alive   

                                          author  \
0                                  Darrell Lucus   
1                                Daniel J. Flynn   
2                             Consortiumnews.com   
3                                Jessica Purkiss   
4                                 Howard Portnoy   
...                                          ...   
20795                              Jerome Hudson   
20796                           Benjamin Hoffman   
20797  Michael J. de la Merced and Rachel Abrams   
20798                                Alex Ansary   
20799                              David Swanson   

                                                    text  label  \
0      House Dem Aide: We Didn’t Even See Comey’s Let...      1   
1      Ever get the feeling your life circles the rou...      0   
2      Why the Truth Might Get You Fired October 29, ...      1   
3      Videos 15 Civilians Killed In Single US Airstr...      1   
4      Print \nAn Iranian woman has been sentenced to...      1   
...                                                  ...    ...   
20795  Rapper T. I. unloaded on black celebrities who...      0   
20796  When the Green Bay Packers lost to the Washing...      0   
20797  The Macy’s of today grew from the union of sev...      0   
20798  NATO, Russia To Hold Parallel Exercises In Bal...      1   
20799    David Swanson is an author, activist, journa...      1   

                                                 content  
0      Darrell Lucus House Dem Aide: We Didn’t Even S...  
1      Daniel J. Flynn FLYNN: Hillary Clinton, Big Wo...  
2      Consortiumnews.com Why the Truth Might Get You...  
3      Jessica Purkiss 15 Civilians Killed In Single ...  
4      Howard Portnoy Iranian woman jailed for fictio...  
...                                                  ...  
20795  Jerome Hudson Rapper T.I.: Trump a ’Poster Chi...  
20796  Benjamin Hoffman N.F.L. Playoffs: Schedule, Ma...  
20797  Michael J. de la Merced and Rachel Abrams Macy...  
20798  Alex Ansary NATO, Russia To Hold Parallel Exer...  
20799            David Swanson What Keeps the F-35 Alive  

[20800 rows x 6 columns]

In [22]:
x=dataset.drop(['label'],axis=1)

In [23]:
x#input dataset

id                                              title  \
0          0  House Dem Aide: We Didn’t Even See Comey’s Let...   
1          1  FLYNN: Hillary Clinton, Big Woman on Campus - ...   
2          2                  Why the Truth Might Get You Fired   
3          3  15 Civilians Killed In Single US Airstrike Hav...   
4          4  Iranian woman jailed for fictional unpublished...   
...      ...                                                ...   
20795  20795  Rapper T.I.: Trump a ’Poster Child For White S...   
20796  20796  N.F.L. Playoffs: Schedule, Matchups and Odds -...   
20797  20797  Macy’s Is Said to Receive Takeover Approach by...   
20798  20798  NATO, Russia To Hold Parallel Exercises In Bal...   
20799  20799                          What Keeps the F-35 Alive   

                                          author  \
0                                  Darrell Lucus   
1                                Daniel J. Flynn   
2                             Consortiumnews.com   
3                                Jessica Purkiss   
4                                 Howard Portnoy   
...                                          ...   
20795                              Jerome Hudson   
20796                           Benjamin Hoffman   
20797  Michael J. de la Merced and Rachel Abrams   
20798                                Alex Ansary   
20799                              David Swanson   

                                                    text  \
0      House Dem Aide: We Didn’t Even See Comey’s Let...   
1      Ever get the feeling your life circles the rou...   
2      Why the Truth Might Get You Fired October 29, ...   
3      Videos 15 Civilians Killed In Single US Airstr...   
4      Print \nAn Iranian woman has been sentenced to...   
...                                                  ...   
20795  Rapper T. I. unloaded on black celebrities who...   
20796  When the Green Bay Packers lost to the Washing...   
20797  The Macy’s of today grew from the union of sev...   
20798  NATO, Russia To Hold Parallel Exercises In Bal...   
20799    David Swanson is an author, activist, journa...   

                                                 content  
0      Darrell Lucus House Dem Aide: We Didn’t Even S...  
1      Daniel J. Flynn FLYNN: Hillary Clinton, Big Wo...  
2      Consortiumnews.com Why the Truth Might Get You...  
3      Jessica Purkiss 15 Civilians Killed In Single ...  
4      Howard Portnoy Iranian woman jailed for fictio...  
...                                                  ...  
20795  Jerome Hudson Rapper T.I.: Trump a ’Poster Chi...  
20796  Benjamin Hoffman N.F.L. Playoffs: Schedule, Ma...  
20797  Michael J. de la Merced and Rachel Abrams Macy...  
20798  Alex Ansary NATO, Russia To Hold Parallel Exer...  
20799            David Swanson What Keeps the F-35 Alive  

[20800 rows x 5 columns]

In [24]:
y=dataset['label']

In [25]:
y.head()#output 

0    1
1    0
2    1
3    1
4    1
Name: label, dtype: int64

#### Stemming:Reduces word to root word like running to run, written to write ,  acting,actor,actress-act

#### Making a function to apply to dataset which will delete all stopwords and and return important text

In [26]:
ps=PorterStemmer()

In [27]:
def stemming(cont):
    stemmed_content=re.sub('[^a-zA-Z]', ' ',cont)#WE get rid of numbers and special character
    #it replaces other then alphabets to spaces
    stemmed_content=stemmed_content.lower()#making all to lowercase
    stemmed_content=stemmed_content.split()#converting to list
    stemmed_content=[ps.stem(word) for word in stemmed_content if not word in stopwords.words('english')]#not to take words that are in stopwords
    stemmed_content=' '.join(stemmed_content)
    return stemmed_content

In [28]:
stemming("This is a sample text with numbers 123 and special characters! Let's see how it works.")

'sampl text number special charact let see work'

In [29]:
dataset['content']=dataset['content'].apply(stemming)

In [30]:
X=dataset['content'].values
Y=dataset['label'].values

In [35]:
print(X)

['darrel lucu hous dem aid even see comey letter jason chaffetz tweet'
 'daniel j flynn flynn hillari clinton big woman campu breitbart'
 'consortiumnew com truth might get fire' ...
 'michael j de la merc rachel abram maci said receiv takeov approach hudson bay new york time'
 'alex ansari nato russia hold parallel exercis balkan'
 'david swanson keep f aliv']


In [36]:
print(Y)

[1 0 1 ... 0 1 1]


#### Encoding text to numerical data using tfifvectorizer()

In [31]:
vec=TfidfVectorizer()#Term Frequency-Inverse Document  Frequency

#### It convert words to matrix . It basically finds repeating words and mark its importance

In [32]:
vec.fit(X)

TfidfVectorizer()

In [37]:
X=vec.transform(X)

In [38]:
print(X)

  (0, 267)	0.2701012497770876
  (0, 2483)	0.36765196867972083
  (0, 2959)	0.24684501285337127
  (0, 3600)	0.3598939188262558
  (0, 3792)	0.27053324808454915
  (0, 4973)	0.23331696690935097
  (0, 7005)	0.2187416908935914
  (0, 7692)	0.24785219520671598
  (0, 8630)	0.2921251408704368
  (0, 8909)	0.36359638063260746
  (0, 13473)	0.2565896679337956
  (0, 15686)	0.2848506356272864
  (1, 1497)	0.2939891562094648
  (1, 1894)	0.15521974226349364
  (1, 2223)	0.3827320386859759
  (1, 2813)	0.19094574062359204
  (1, 3568)	0.26373768806048464
  (1, 5503)	0.7143299355715573
  (1, 6816)	0.1904660198296849
  (1, 16799)	0.30071745655510157
  (2, 2943)	0.3179886800654691
  (2, 3103)	0.46097489583229645
  (2, 5389)	0.3866530551182615
  (2, 5968)	0.3474613386728292
  (2, 9620)	0.49351492943649944
  :	:
  (20797, 3643)	0.2115550061362374
  (20797, 7042)	0.21799048897828685
  (20797, 8364)	0.22322585870464115
  (20797, 8988)	0.36160868928090795
  (20797, 9518)	0.29542040034203126
  (20797, 9588)	0.17455348

### Train,test_split

In [41]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=2)

#### LOGISTIC REGRESSION

In [42]:
lg=LogisticRegression()

In [43]:
lg.fit(X_train,y_train)

LogisticRegression()

#### Evaluation

#### Accuracy test on training data

In [44]:
X_train_pred=lg.predict(X_train)

In [45]:
train_acc=accuracy_score(X_train_pred,y_train)

In [46]:
train_acc

0.9871995192307692

In [50]:
Xtest_pred=lg.predict(X_test)

In [51]:
accuracy_score(Xtest_pred,y_test)

0.9774038461538461

#### Making predictive system

In [59]:
X_new=X_test[6]

In [60]:
print(X_new)

  (0, 919)	0.26781002972130336
  (0, 1731)	0.3075650098229431
  (0, 1825)	0.2740244377245676
  (0, 1894)	0.11674670065135907
  (0, 3493)	0.24296459571195272
  (0, 5022)	0.24130960904374057
  (0, 5167)	0.45236218497875413
  (0, 5544)	0.22159700452199232
  (0, 5749)	0.2333163716858574
  (0, 9310)	0.23017748071475028
  (0, 10992)	0.25147070160022716
  (0, 14711)	0.27500228188923115
  (0, 15050)	0.2694834111534771
  (0, 15142)	0.23464907311239305


In [62]:
predict=lg.predict(X_new)

In [63]:
if predict==0:
    print("Real news")
else:
    print("False news")

Real news


In [67]:
y_test

15920    1
2420     1
3589     1
2490     1
13482    1
        ..
18172    0
8112     0
852      1
9631     1
2500     1
Name: label, Length: 4160, dtype: int64

ValueError: Iterable over raw text documents expected, string object received.